### PR TITLE
add block-version field to propose tx response

### LINKED ISSUE #1718: Add a `block_version` field to `ProposeTxResponse`
Mobile SDK team reports:

> we're thinking about how we manage the `BlockVersion` state in the SDKs.
>
> We can cache the result from the message `CheckKeyImagesResponse` (https://github.com/mobilecoinfoundation/mobilecoin/blob/master/fog/api/proto/ledger.proto#L124)  which happens in a balance check. If consensus restarts the value may be incorrect and transaction would get rejected. We could then do a network round-trip to `CheckKeyImagesResponse` , re-build, and re-submit. Now if the current `BlockVersion` came back in the error message or as part of the `ProposeTxResponse` we'd be able to re-build without an extra network call, does it make sense to add that ?

---

Suggested resolution:

Add a `block_version` field to this response structure:

https://github.com/mobilecoinfoundation/mobilecoin/blob/50a8e69dfe03f2572b79d2e23c75066069fb799f/consensus/api/proto/consensus_common.proto#L87

and make consensus service just always tell the clients what the current rules are, whether or not there is an error

### PR DESCRIPTION
**Pull request by garbageslam**
Fixes #1718

### CODE REVIEW

**Review by voloshyn — APPROVED** [review]
LGTM

File: consensus/service/src/api/client_api_service.rs [hunk 1]
```diff
@@ -188,6 +188,14 @@ impl ConsensusClientApi for ClientApiService {
...
             let num_blocks = self.ledger.num_blocks().map_err(ConsensusGrpcError::from)?;
             response.set_block_count(num_blocks);
+
+            // This is okay because it is invalid to start consensus without an origin block
+            // anyways
+            let block = self
+                .ledger
+                .get_block(num_blocks - 1)
+                .map_err(ConsensusGrpcError::from)?;
+            response.set_block_version(block.version);
```
Comment (eranrund) [inline consensus/service/src/api/client_api_service.rs:198]: This seems wrong to me. We could be on version 3 even if we haven't produced a version 3 block yet (e.g. when we are about to transition from 2 to 3).

Comment (garbageslam) [reply consensus/service/src/api/client_api_service.rs:198]: i see, i guess you are right. will fix

**Comment by jcape** [comment]
I think this is slightly more involved, we should be keeping the block version state in the enclave, can calling into it to get it, since it's the arbiter of truth in both the config version of reality (0 -> 1) and the "mint tx" version of reality (1 -> 2).

**Comment by eranrund** [comment]
> I think this is slightly more involved, we should be keeping the block version state in the enclave, can calling into it to get it, since it's the arbiter of truth in both the config version of reality (0 -> 1) and the "mint tx" version of reality (1 -> 2).

I am not sure what that gets us, since untrusted feeds whatever block version it wants into the enclave (from the `Config` struct). What do you think getting it from the enclave gives us in comparison to getting it from the configuration struct?

**Comment by garbageslam** [comment]
> both the config version of reality (0 -> 1) and the "mint tx" version of reality (1 -> 2).

i am not sure what this means, to my knowledge, mint tx validation takes place outside the enclave. if it was moved in, then that was a scope creep for the enclave and IMO should be reverted

File: consensus/service/src/api/client_api_service.rs [hunk 2]
```diff
@@ -188,6 +192,7 @@ impl ConsensusClientApi for ClientApiService {
         result = result.and_then(|mut response| {
             let num_blocks = self.ledger.num_blocks().map_err(ConsensusGrpcError::from)?;
             response.set_block_count(num_blocks);
+            response.set_block_version(*self.config.block_version);
```
Comment (iamalwaysuncomfortable) [inline consensus/service/src/api/client_api_service.rs:195]: If this is approved @christian-oudard & FS should be notified to update their protos responses to include it as well

**Review by iamalwaysuncomfortable — APPROVED** [review]
Looks good, just recommend tapping full-service team on the shoulder to update their protos too

**Review by sugargoat — APPROVED** [review]

**Review by eranrund — APPROVED** [review]

**Pull request merged**